src/build_system.ml

```diff
@@ -704,58 +713,63 @@ and static_deps t build =
 and start_rule t _rule =
   t.hook Rule_started
 
-and setup_copy_rules t ~ctx_dir ~non_target_source_files =
-  Path.Source.Set.iter non_target_source_files ~f:(fun path ->
-    let ctx_path = Path.append_source ctx_dir path in
-    let build = Build.copy ~src:(Path.source path) ~dst:ctx_path in
-    compile_rule t (Pre_rule.make build ~context:None ~env:None
-                      ~info:Source_file_copy))
+and create_copy_rules ~ctx_dir ~non_target_source_files =
+  List.map
+    (Path.Source.Set.to_list non_target_source_files)
+    ~f:(fun path ->
+      let ctx_path = Path.append_source ctx_dir path in
+      let build = Build.copy ~src:(Path.source path) ~dst:ctx_path in
+      (Pre_rule.make build ~context:None ~env:None
+         ~info:Source_file_copy))
+
+and compile_rules ~dir t rules =
+  List.map rules ~f:(compile_rule t)
```

Comment:
`compile_rule` can be just done in the `concat_map`

Reply:
done